File: src/VBNetFx/ADDataTypes/ADDBNullable.vb
Public Interface IADDBNullableValue
    ReadOnly Property IsNull() As Boolean
    ReadOnly Property HasValue() As Boolean
    ReadOnly Property Value() As Object
End Interface

Public Module ADDBNullable
    Public Function FromObject(Of PrimType As IComparable)(ByVal value As Object) As ADDBNullable(Of PrimType)
        Dim locNullable As New ADDBNullable(Of PrimType)
        If (value Is Nothing) Then
            Return locNullable
        ElseIf (TypeOf value Is DBNull) Then
            Return locNullable
        End If
        If TypeOf value Is PrimType Then
            locNullable.myValue = CType(value, PrimType)
            locNullable.myNotNull = True
            Return locNullable
        End If

        'Ein Versuch ist es Wert!
        Try
            locNullable.myValue = CType(value, PrimType)
            locNullable.myNotNull = True
            Return locNullable
        Catch ex As Exception
            Dim up As New InvalidCastException("Object is not of the correct type!")
            Throw up
        End Try
    End Function

    ''' <summary>
    ''' Wandelt in den gekapselten primitiven Datentypen um, 
    ''' </summary>
    ''' <typeparam name="PrimType">Der zugrundeliegende Basistyp</typeparam>
    ''' <param name="value">In seinen primitiven Typ zu konvertierender ADDBNullable</param>
    ''' <returns>Object mit geboxtem primitivem Typ</returns>
    ''' <remarks></remarks>
    Public Function ToObject(Of PrimType As IComparable)(ByVal value As ADDBNullable(Of PrimType)) As Object
        If Not value.HasValue Then
            Return Nothing
        End If
        Return CType(value.Value, PrimType)
    End Function
End Module

<CLSCompliant(True), Serializable()> _
Public Structure ADDBNullable(Of PrimType As IComparable)
    Implements IComparable, IADDBNullableValue, System.Runtime.Serialization.ISerializable

    Friend myValue As PrimType
    Friend myNotNull As Boolean

    Sub New(ByVal Value As PrimType)
        'Sonderbehandlung String, damit "" impliziet in DBNull umgewandelt werden kann
        If GetType(PrimType) Is GetType(String) Then
            If Convert.ToString(Value) = "" Then
                myNotNull = False
                myValue = Nothing
                Return
            End If
        End If

        If Value Is Nothing Then
            myNotNull = False
        Else
            myNotNull = True
        End If
        myValue = Value
    End Sub

    Public ReadOnly Property IsNull() As Boolean Implements IADDBNullableValue.IsNull
        Get
            Return Not myNotNull
        End Get
    End Property

    Public ReadOnly Property HasValue() As Boolean Implements IADDBNullableValue.HasValue
        Get
            Return myNotNull
        End Get
    End Property

    <Xml.Serialization.XmlIgnore()> _
    Public ReadOnly Property Value() As Object Implements IADDBNullableValue.Value
        Get
            If IsNull Then
                Return DBNull.Value
            Else
                Return myValue
            End If
        End Get
    End Property

    Public ReadOnly Property TypedValue() As PrimType
        Get
            If IsNull Then
                Dim up As New InvalidCastException("Can't cast DBNull to its native type")
                Throw up
            Else
                Return myValue
            End If
        End Get
    End Property

    Public Shared Widening Operator CType(ByVal value As PrimType) As ADDBNullable(Of PrimType)
        Return New ADDBNullable(Of PrimType)(value)
    End Operator

    Public Shared Widening Operator CType(ByVal value As DBNull) As ADDBNullable(Of PrimType)
        Return New ADDBNullable(Of PrimType)()
    End Operator

    Public Shared Widening Operator CType(ByVal value As ADDBNullable(Of PrimType)) As PrimType
        If value.IsNull Then
            Return Nothing
        End If
        Return value.myValue
    End Operator

    Public Shared Operator =(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As PrimType) As Boolean
        Return value1.CompareTo(value2) = 0
    End Operator

    Public Shared Operator =(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As ADDBNullable(Of PrimType)) As Boolean
        Return value1.CompareTo(value2.Value) = 0
    End Operator

    Public Shared Operator <>(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As PrimType) As Boolean
        Return Not value1.CompareTo(value2) = 0
    End Operator

    Public Shared Operator <>(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As ADDBNullable(Of PrimType)) As Boolean
        Return Not value1.CompareTo(value2.Value) = 0
    End Operator

    Public Shared Operator >(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As PrimType) As Boolean
        Return value1.CompareTo(value2) = 1
    End Operator

    Public Shared Operator >(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As ADDBNullable(Of PrimType)) As Boolean
        Return value1.CompareTo(value2.Value) = 1
    End Operator

    Public Shared Operator <(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As PrimType) As Boolean
        Return value1.CompareTo(value2) = -1
    End Operator

    Public Shared Operator <(ByVal value1 As ADDBNullable(Of PrimType), ByVal value2 As ADDBNullable(Of PrimType)) As Boolean
        Return value1.CompareTo(value2.Value) = -1
    End Operator

    Public Function CompareTo(ByVal obj As Object) As Integer Implements System.IComparable.CompareTo
        If obj.GetType Is GetType(DBNull) Then
            If IsNull Then
                Return 0
            Else
                Return 1
            End If
        End If
        Return myValue.CompareTo(obj)
    End Function

    Public Overrides Function ToString() As String
        If IsNull Then
            Return ""
        Else
            Return myValue.ToString
        End If
    End Function

    Public Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext) Implements System.Runtime.Serialization.ISerializable.GetObjectData

        If IsNull Then
            info.SetType(GetType(DBNull))
            info.AddValue("ADDBNullable", DBNull.Value)
        Else
            info.SetType(GetType(PrimType))
            info.AddValue("ADDBnullable", Me.TypedValue)
        End If
    End Sub
End Structure

File: src/MixNetFx/ADDataTypes/ADDBNullable.cs
using System;
using System.Runtime.Serialization;
using System.Xml.Serialization;

namespace ActiveDev
{
    public interface IADDBNullableValue
    {
        bool IsNull { get; }
        bool HasValue { get; }
        object Value { get; }
    }

    public static class ADDBNullable
    {
        public static ADDBNullable<PrimType> FromObject<PrimType>(object value)
            where PrimType : IComparable
        {
            var locNullable = new ADDBNullable<PrimType>();
            if (value == null || value is DBNull)
            {
                return locNullable;
            }

            if (value is PrimType typedValue)
            {
                locNullable.myValue = typedValue;
                locNullable.myNotNull = true;
                return locNullable;
            }

            try
            {
                locNullable.myValue = (PrimType)Convert.ChangeType(value, typeof(PrimType));
                locNullable.myNotNull = true;
                return locNullable;
            }
            catch (Exception)
            {
                throw new InvalidCastException("Object is not of the correct type!");
            }
        }

        public static object ToObject<PrimType>(ADDBNullable<PrimType> value)
            where PrimType : IComparable
        {
            if (!value.HasValue)
            {
                return null;
            }

            return (PrimType)value.Value;
        }
    }

    [CLSCompliant(true)]
    [Serializable]
    public struct ADDBNullable<PrimType> : IComparable, IADDBNullableValue, ISerializable
        where PrimType : IComparable
    {
        internal PrimType myValue;
        internal bool myNotNull;

        public ADDBNullable(PrimType value)
        {
            if (typeof(PrimType) == typeof(string) && Convert.ToString(value) == string.Empty)
            {
                myNotNull = false;
                myValue = default(PrimType);
                return;
            }

            myNotNull = !ReferenceEquals(value, null);
            myValue = value;
        }

        public bool IsNull
        {
            get { return !myNotNull; }
        }

        public bool HasValue
        {
            get { return myNotNull; }
        }

        [XmlIgnore]
        public object Value
        {
            get { return IsNull ? (object)DBNull.Value : myValue; }
        }

        public PrimType TypedValue
        {
            get
            {
                if (IsNull)
                {
                    throw new InvalidCastException("Can't cast DBNull to its native type");
                }

                return myValue;
            }
        }

        public static implicit operator ADDBNullable<PrimType>(PrimType value)
        {
            return new ADDBNullable<PrimType>(value);
        }

        public static implicit operator ADDBNullable<PrimType>(DBNull value)
        {
            return new ADDBNullable<PrimType>();
        }

        public static implicit operator PrimType(ADDBNullable<PrimType> value)
        {
            if (value.IsNull)
            {
                return default(PrimType);
            }

            return value.myValue;
        }

        public static bool operator ==(ADDBNullable<PrimType> value1, PrimType value2)
        {
            return value1.CompareTo(value2) == 0;
        }

        public static bool operator ==(ADDBNullable<PrimType> value1, ADDBNullable<PrimType> value2)
        {
            return value1.CompareTo(value2.Value) == 0;
        }

        public static bool operator !=(ADDBNullable<PrimType> value1, PrimType value2)
        {
            return value1.CompareTo(value2) != 0;
        }

        public static bool operator !=(ADDBNullable<PrimType> value1, ADDBNullable<PrimType> value2)
        {
            return value1.CompareTo(value2.Value) != 0;
        }

        public static bool operator >(ADDBNullable<PrimType> value1, PrimType value2)
        {
            return value1.CompareTo(value2) == 1;
        }

        public static bool operator >(ADDBNullable<PrimType> value1, ADDBNullable<PrimType> value2)
        {
            return value1.CompareTo(value2.Value) == 1;
        }

        public static bool operator <(ADDBNullable<PrimType> value1, PrimType value2)
        {
            return value1.CompareTo(value2) == -1;
        }

        public static bool operator <(ADDBNullable<PrimType> value1, ADDBNullable<PrimType> value2)
        {
            return value1.CompareTo(value2.Value) == -1;
        }

        public int CompareTo(object obj)
        {
            if (obj.GetType() == typeof(DBNull))
            {
                return IsNull ? 0 : 1;
            }

            return myValue.CompareTo(obj);
        }

        public override bool Equals(object obj)
        {
            if (obj is ADDBNullable<PrimType> nullableValue)
            {
                return this == nullableValue;
            }

            if (obj is PrimType typedValue)
            {
                return this == typedValue;
            }

            return false;
        }

        public override int GetHashCode()
        {
            return IsNull ? 0 : myValue.GetHashCode();
        }

        public override string ToString()
        {
            return IsNull ? string.Empty : myValue.ToString();
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            if (IsNull)
            {
                info.SetType(typeof(DBNull));
                info.AddValue("ADDBNullable", DBNull.Value);
            }
            else
            {
                info.SetType(typeof(PrimType));
                info.AddValue("ADDBnullable", TypedValue);
            }
        }
    }
}
